### ISSUE TITLE
treat Dataset_Xxx as a standard HLT path for run3 data

### ISSUE THREAD

**Issue by xiezhen** [xiezhen/brilws#3]
In Run-3, a dataset "Xxx" is associated to an HLT path called "Dataset_Xxx", and its prescales can be queried like for other HLT paths.

For example, to query the luminosity recorded by the "PhysicsHLTPhysics2" dataset, you can look for the prescale of the path "Dataset_PhysicsHLTPhysics2".

So, to check the amount of luminosity recorded by the "HLT_Physics_v8" path into the "PhysicsHLTPhysics2" dataset, you could
  - query the prescale of the HLT_Physics_v8 path
  - query the prescale of the Dataset_PhysicsHLTPhysics2 path
  - multiply them

(there is a special caveat with the HLT_Physics* paths: they don't have an L1 seed so brilcalc fails, and they have an implicit additional prescale of 107, but these issues are beside the point of this request)

Of course this makes sense only if the path is actually part of the dataset...

**Comment by xiezhen** [comment]
select distinct(path) from table (cms_hlt_gdr.get_prescales_byidx(362728,8));
gives back only HLT_* paths.

Function cms_hlt_gdr.get_prescales_byidx is our entry point function in cms_hlt_gdr to prescale values, 8 is the first prescale index of this run. 

Where 362728 is a pp run with Datase_* path ( as seen from confdbv3).

**Comment by xiezhen** [comment]
Hi Sandro,

   I’ve tried the following query: select * from  ( cms_hlt_gdr.get_ugtprescales(362107,23)) where PATH='Dataset_EphemeralHLTPhysics2’;
   And also PATH like 'Dataset like Dataset_%’; 

   I don’t get any results. I am using acccount cms_runinfo_r for reading. Could you make sure that this function is open to cms_runinfo_r for read accecss?
  I also noticed that this is a new function. Does it mean that it should replace the old cms_hlt_gdr.get_prescales_byidx ?
 
thanks,
Zhen

**Comment by xiezhen** [comment]
Hi Zhen,
   the returned PATH field content has double quotes around the path name, so your select query must be adapted accordingly:

select * from  cms_hlt_gdr.get_ugtprescales(362107,23) where PATH='"Dataset_EphemeralHLTPhysics2”'

Another query could be:

select * from  cms_hlt_gdr.get_ugtprescales(362107,23) where INSTR(PATH,'Dataset_EphemeralHLTPhysics2')>0

**Comment by xiezhen** [comment]
Hi Andrea,

    I don’t know how to get the relationship between a dataset and it’s “children" hltpaths. i.e. between  "PhysicsHLTPhysics2”  and "HLT_Physics_v8” in your example.

1. Suppose such knowledge is not required from BRIL side, I’d require the user to be correct  in dataset, hltpath relationship of the same menu and with no typing errors, when providing the —hltpath and —dataset  arguments: i.e.
    brilcalc lumi -r runnumber --hltpath HLT_Physics_v8 --dataset PhysicsHLTPhysics2

Assuming the client himself guarantees the relationship between  PhysicsHLTPhysics2  and HLT_Physics_v8. I can query prescale of  both for each lumi section ( as Sandro showed possible), then I’d multiply the 2 prescale factors of —dataset and —hltpath, as you described. Then goes further to search for its l1 seed for —hltpath and do prescale algebra as before.

2. Suppose that the client cannot guarantee the correctness of —dataset and —hltpath arguments, then the BRIL database has to store all  the possible dataset to hltpath relationships in advance. In that case, I miss the queries how to do the following :
    —  given a hltconfig id get all the dataset_path and hlt_path associated with each dataset, e.g.
          select dataset_pathid, dataset_pathname, hlt_pathid,hlt_pathname where hltconfigid=1234;

For querying the hltpath to l1seed relationship, it’s quite a complex query which uses the relationship name “L1SeedsLogicalExpression” . This cannot be used in this case. I’d need a new query. 

So is scenario 2 required?

thanks,
Zhen

**Comment by xiezhen** [comment]
Hi Zhen,
in principle we could leave it to the user to specify the correct dataset <-> path association (as in your case 1.)… but it would be easy for people to make a mistake, get the wrong prescales, and effectively the wrong luminosity :-(

So it would indeed be better to extract and store the dataset <-> path association as in your case 2.), similarly to what is done for the HLT path <-> L1 trigger association.

I can explain how the information is encoded within the HLT menu.
Hopefully that is enough information that Sandro can write the corresponding query !

Each “dataset path” is composed by three parts:

the HLTDatasetPathBeginSequence (should be identical for all datasets, and we can ignore it here)
a TriggerResultsFilter that lists what paths are part of the dataset
an HLTPrescaler that applies the prescale to the dataset


For example, the “ParkingDoubleMuonLowMass0” dataset is defined by the Dataset_ParkingDoubleMuonLowMass0 path:

fragment.Dataset_ParkingDoubleMuonLowMass0 = cms.Path( fragment.HLTDatasetPathBeginSequence + fragment.hltDatasetParkingDoubleMuonLowMass + fragment.hltPreDatasetParkingDoubleMuonLowMass0 )
and the hltDatasetParkingDoubleMuonLowMass lists the corresponding HLT paths:

fragment.hltDatasetParkingDoubleElectronLowMass = cms.EDFilter( "TriggerResultsFilter",
    usePathStatus = cms.bool( True ),
    hltResults = cms.InputTag( "" ),
    l1tResults = cms.InputTag( "" ),
    l1tIgnoreMaskAndPrescale = cms.bool( False ),
    throw = cms.bool( True ),
    triggerConditions = cms.vstring( 'HLT_DoubleEle10_eta1p22_mMax6_dz0p8_v1',
      'HLT_DoubleEle10_eta1p22_mMax6_trkHits10_v1',
      'HLT_DoubleEle10_eta1p22_mMax6_v2',
      'HLT_DoubleEle4_eta1p22_mMax6_dz0p8_v1',
      ...
      'HLT_SingleEle8_SingleEGL1_v1',
      'HLT_SingleEle8_v1' )
)


So

the list of all datasets can be extracted from the list of all path, looking for those whose name starts with Dataset_
for each dataset, take the corresponding path
look for a TriggerResultsFilter module on the path (there should be only one !)
extract the triggerConditions parameter from that module, which should be of type cms.vstring and contain a list of HLT paths
those are the paths that should be associated to the dataset


Sandro, is this feasible via a query ?

Ciao,
.Andrea

**Comment by xiezhen** [comment]
Hi,
   as Andrea suggests, the Dataset structure can surely be extracted from the ConfDB. I’ll give a look and let you know.

ciao
sandro

**Comment by xiezhen** [comment]
implemented with release 3.7.0 https://cms-talk.web.cern.ch/t/bril-worksuite-new-release-3-7-0/20099
Note: option --dataset has no searching capability due to lack of knowledge about dataset and hltpath relationships.

**Comment by fwyzard** [comment]
hi @xiezhen, I've given it a try and it seems to work correctly in my test:
```
fwyzard@lxplus8s16:~$ brilcalc lumi -r 362616 --hltpath HLT_EphemeralZeroBias_v3 
#Data tag : 23v1 , Norm tag: onlineresult
+-------------+-------------------+------+--------------------------+----------------+----------------+
| run:fill    | time              | ncms | hltpath                  | delivered(/ub) | recorded(/ub)  |
+-------------+-------------------+------+--------------------------+----------------+----------------+
| 362616:8474 | 11/24/22 09:14:12 | 285  | HLT_EphemeralZeroBias_v3 | 9917.938377291 | 9044.907995677 |
+-------------+-------------------+------+--------------------------+----------------+----------------+
#Summary: 
+--------------------------+-------+------+------+-------------------+------------------+
| hltpath                  | nfill | nrun | ncms | totdelivered(/ub) | totrecorded(/ub) |
+--------------------------+-------+------+------+-------------------+------------------+
| HLT_EphemeralZeroBias_v3 | 1     | 1    | 285  | 9917.938377291    | 9044.907995677   |
+--------------------------+-------+------+------+-------------------+------------------+
#Sum delivered : 9917.938377291
#Sum recorded : 9044.907995677
```
```
fwyzard@lxplus8s16:~$ brilcalc lumi -r 362616 --hltpath HLT_EphemeralZeroBias_v3 --dataset Dataset_EphemeralZeroBias0
#Data tag : 23v1 , Norm tag: onlineresult
+-------------+-------------------+------+--------------------------+----------------+----------------+
| run:fill    | time              | ncms | hltpath                  | delivered(/ub) | recorded(/ub)  |
+-------------+-------------------+------+--------------------------+----------------+----------------+
| 362616:8474 | 11/24/22 09:14:12 | 285  | HLT_EphemeralZeroBias_v3 | 1101.993153032 | 1004.989777297 |
+-------------+-------------------+------+--------------------------+----------------+----------------+
#Summary: 
+--------------------------+-------+------+------+-------------------+------------------+
| hltpath                  | nfill | nrun | ncms | totdelivered(/ub) | totrecorded(/ub) |
+--------------------------+-------+------+------+-------------------+------------------+
| HLT_EphemeralZeroBias_v3 | 1     | 1    | 285  | 1101.993153032    | 1004.989777297   |
+--------------------------+-------+------+------+-------------------+------------------+
#Sum delivered : 1101.993153032
#Sum recorded : 1004.989777297
```

I found it a bit confusing that I need to specify `Dataset_EphemeralZeroBias0` instead of simply `EphemeralZeroBias0`, since the name that we normally use for the dataset is the latter.

Could you make `brilcalc` automatically add `Dataset_` in front of the dataset name ?

**Comment by xiezhen** [comment]
OK, we can do that. I didn't know about these conventions.

**Comment by xiezhen** [comment]
It's fixed with release 3.7.1

**Comment by fwyzard** [comment]
Nice !
